### PR TITLE
EIP-7951: Precompile for secp256r1 Curve Support

### LINKED ISSUE #2363: EIP-7951: Precompile for secp256r1 Curve Support

### PR DESCRIPTION
**Pull request by lorenzogentile404**
> [!NOTE]
> Add full P-256 (secp256r1) verify precompile tracing, OOB/MMU handling, and tests across the tracer.
> 
> - **Precompile support (Osaka/post-Cancun)**:
>   - Add `PRC_P256_VERIFY` throughout: `ModuleName`, counters (`ZkCounter`, `Hub`), and `EcData`/`EcDataOperation` (new phases, sizes, range checks, R1 curve membership, result handling).
>   - Implement new OOB call `P256VerifyOobCall` and refactor OOB packages from `bls` to `postCancun` (including common fixed-size/gas-cost base); gate call-data extraction via `getCdxFilter`.
>   - Update MMU plumbing: rename BLS helpers to generic post-cancun helpers and route P256 paths for call-data/return-data transfers; add helper to select EC vs BLS module.
>   - Extend scenario/flag maps, phases, CT limits, and validators for P256 in `PrecompileScenarioFragment`.
> - **Utilities/Costs**:
>   - Add P256 gas/return-size to `PrecompileUtils`; switch cost/return-capacity logic to enum-based dispatch.
> - **Tests**:
>   - Add comprehensive P256 verify tests (success/failure, gas stipend, sizes, curve/encoding edge cases; include large test vectors JSON).
>   - Broaden post-Cancun precompile size tests and low-gas stipend tests to cover P256; rename tests accordingly.
> - **Misc**:
>   - Minor README formatting tweaks; internal API tweaks (`fillAndValidateRow`, cleanup of selector fields).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b5775ceca93b825bae3fe47d63fd08f856ad1931. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/blake2fmodexpdata/BlakeModexpDataOperation.java:1]: not related to this PR and will create conflict with MODEXP pr

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/ecdata/EcDataOperation.java [hunk 1]
```diff
@@ -729,7 +792,11 @@ void trace(Trace.Ecdata trace, final int stamp, final long previousId) {
           .isEcmulData(precompileFlag == PRC_ECMUL && isData)
           .isEcmulResult(precompileFlag == PRC_ECMUL && !isData)
           .isEcpairingData(precompileFlag == PRC_ECPAIRING && isData)
-          .isEcpairingResult(precompileFlag == PRC_ECPAIRING && !isData)
+          .isEcpairingResult(precompileFlag == PRC_ECPAIRING && !isData);
+      if (precompileFlag == PRC_P256_VERIFY) {
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/ecdata/EcDataOperation.java:797]: write like the other, it seems it has a different behavior, but no

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java [hunk 2]
```diff
@@ -574,57 +574,78 @@ public static MmuCall callDataExtractionForEcpairing(
...
+            .sourceRamBytes(Optional.of(subsection.rawCallerMemory()))
+            .targetId(precompileContextNumber)
+            .exoBytes(Optional.of(subsection.extractCallData()))
+            .sourceOffset(EWord.of(subsection.callDataOffset()))
+            .size(subsection.callDataSize())
+            .referenceSize(subsection.callDataSize())
+            // constant
+            .successBit(successBit)
+            .phase(subsection.flag().dataPhase());
+    if (subsection.flag() == PrecompileScenarioFragment.PrecompileFlag.PRC_P256_VERIFY) {
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: I'd prefer mmuCall.setEcData(prc == BLA).
easier to read, and you can return directly the mmuCall without doing in two steps

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java [hunk 3]
```diff
@@ -574,57 +574,78 @@ public static MmuCall callDataExtractionForEcpairing(
...
+        new MmuCall(hub, MMU_INST_EXO_TO_RAM_TRANSPLANTS)
+            .sourceId(precompileContextNumber)
+            .exoBytes(Optional.of(subsection.returnDataRange.extract()))
+            .targetId(precompileContextNumber)
+            .targetRamBytes(Optional.of(Bytes.EMPTY))
+            .size(expectedReturnDataSize)
+            .phase(subsection.flag().resultPhase())
+            .successBit(successBit);
+
+    if (subsection.flag() == PrecompileScenarioFragment.PrecompileFlag.PRC_P256_VERIFY) {
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: same

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java [hunk 4]
```diff
@@ -574,57 +574,78 @@ public static MmuCall callDataExtractionForEcpairing(
...
+            .sourceId(precompileContextNumber)
+            .sourceRamBytes(Optional.of(subsection.returnDataRange.extract()))
+            .targetId(hub.currentFrame().contextNumber())
+            .targetRamBytes(Optional.of(subsection.rawCallerMemory()))
+            .size(returnDataSize)
+            .referenceOffset(subsection.returnAtOffset())
+            .referenceSize(subsection.returnAtCapacity());
+
+    // TODO: do we need this?
+    if (subsection.flag() == PrecompileScenarioFragment.PrecompileFlag.PRC_P256_VERIFY) {
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: same (and yes we need the exoflag :)

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/EllipticCurvePrecompileSubsection.java [hunk 5]
```diff
@@ -117,6 +119,12 @@ public void resolveAtContextReEntry(Hub hub, CallFrame callFrame) {
           PRC_BLS_MAP_FP2_TO_G2 -> {
         // Note that BLS sanity checks are computed in BlsOperation
       }
+        // TODO: is this correct?
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/EllipticCurvePrecompileSubsection.java]: can you solve the TODO before merging plz ?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/EllipticCurvePrecompileSubsection.java [hunk 6]
```diff
@@ -128,6 +136,7 @@ public void resolveAtContextReEntry(Hub hub, CallFrame callFrame) {
     //    hubSuccess ≡ true
 
     // ECRECOVER can only be FAILURE_KNOWN_TO_HUB or some form of SUCCESS_XXXX_REVERT
+    // TODO: what about P256_VERIFY?
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/EllipticCurvePrecompileSubsection.java]: same

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/instructions/ModexpZero.java [hunk 7]
```diff
@@ -16,8 +16,8 @@
 package net.consensys.linea.zktracer.module.mmu.instructions;
 
 import static net.consensys.linea.zktracer.Trace.MMIO_INST_LIMB_VANISHES;
-import static net.consensys.linea.zktracer.Trace.Mmu.NB_MICRO_ROWS_TOT_MODEXP_ZERO;
 import static net.consensys.linea.zktracer.Trace.Mmu.NB_PP_ROWS_MODEXP_ZERO;
+import static net.consensys.linea.zktracer.TraceOsaka.Mmu.NB_MICRO_ROWS_TOT_MODEXP_ZERO;
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/instructions/ModexpZero.java]: not related to this PR ...

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/instructions/ModexpData.java [hunk 8]
```diff
@@ -20,8 +20,8 @@
 import static net.consensys.linea.zktracer.Trace.MMIO_INST_RAM_TO_LIMB_ONE_SOURCE;
 import static net.consensys.linea.zktracer.Trace.MMIO_INST_RAM_TO_LIMB_TRANSPLANT;
 import static net.consensys.linea.zktracer.Trace.MMIO_INST_RAM_TO_LIMB_TWO_SOURCE;
-import static net.consensys.linea.zktracer.Trace.Mmu.NB_MICRO_ROWS_TOT_MODEXP_DATA;
 import static net.consensys.linea.zktracer.Trace.Mmu.NB_PP_ROWS_MODEXP_DATA;
+import static net.consensys.linea.zktracer.TraceLondon.Mmu.NB_MICRO_ROWS_TOT_MODEXP_DATA;
```
Comment (letypequividelespoubelles) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/instructions/ModexpData.java:25]: not related to this PR ...

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ZkCounter.java [hunk 9]
```diff
@@ -337,7 +339,8 @@ public ZkCounter(LineaL1L2BridgeSharedConfiguration bridgeConfiguration) {
             ecRecoverEffectiveCall,
             ecPairingG2MembershipCalls,
             ecPairingMillerLoops,
-            ecPairingFinalExponentiations);
+            ecPairingFinalExponentiations,
```
[the remark this answers is not in the document]
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/ZkCounter.java:362]: cursor is right

**Review by letypequividelespoubelles — CHANGES_REQUESTED** [review]

File: arithmetization/src/test/resources/p256_verify_test_vectors.json
Comment (letypequividelespoubelles) [inline arithmetization/src/test/resources/p256_verify_test_vectors.json:1]: Is it a copy paste from somewhere ? If yes, wouldn't it be better to point to a specific commit ?

Comment (lorenzogentile404) [reply arithmetization/src/test/resources/p256_verify_test_vectors.json:1]: https://eips.ethereum.org/assets/eip-7951/test-vectors.json it comes from the EIP.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/ecdata/EcDataOperation.java (hunk 1 above)
Comment (lorenzogentile404) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/ecdata/EcDataOperation.java:797]: If we are not dealing with the Osaka trace, then the method for tracing the `P256_VERIFY` flag column does not exist, so I believe it should stay as it is.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java (hunk 2 above)
Comment (lorenzogentile404) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: Done.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java (hunk 3 above)
Comment (lorenzogentile404) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: Done.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java (hunk 4 above)
Comment (lorenzogentile404) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: Done.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/EllipticCurvePrecompileSubsection.java (hunk 5 above)
Comment (lorenzogentile404) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/precompileSubsection/EllipticCurvePrecompileSubsection.java]: You you ask so kindly, yes.

**Review by letypequividelespoubelles — APPROVED** [review]
LGTM

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java (hunk 2 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java]: what I had in mind is : https://github.com/Consensys/linea-tracer/pull/2452/commits/ab9fffd40694d9549041b1ac6c69eb16b333fb11 
I find it cleaner
